# Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('horse_data.csv', na_values = '?', names = ['surgery', 'Age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination', 'abdomen', 'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome', 'surgical lesion', 'type of lesion1', 'type of lesion2', 'type of lesion3', 'cp_data'])
# Добавим сразу названия столбцов в датафрейм согласно описанию и заменим вопросы на НаН
df.head(20)

surgery  Age  Hospital Number  rectal temperature  pulse  \
0       2.0    1           530101                38.5   66.0   
1       1.0    1           534817                39.2   88.0   
2       2.0    1           530334                38.3   40.0   
3       1.0    9          5290409                39.1  164.0   
4       2.0    1           530255                37.3  104.0   
5       2.0    1           528355                 NaN    NaN   
6       1.0    1           526802                37.9   48.0   
7       1.0    1           529607                 NaN   60.0   
8       2.0    1           530051                 NaN   80.0   
9       2.0    9          5299629                38.3   90.0   
10      1.0    1           528548                38.1   66.0   
11      2.0    1           527927                39.1   72.0   
12      1.0    1           528031                37.2   42.0   
13      2.0    9          5291329                38.0   92.0   
14      1.0    1           534917                38.2   76.0   
15      1.0    1           530233                37.6   96.0   
16      1.0    9          5301219                 NaN  128.0   
17      2.0    1           526639                37.5   48.0   
18      1.0    1          5290481                37.6   64.0   
19      2.0    1           532110                39.4  110.0   

    respiratory rate  temperature of extremities  peripheral pulse  \
0               28.0                         3.0               3.0   
1               20.0                         NaN               NaN   
2               24.0                         1.0               1.0   
3               84.0                         4.0               1.0   
4               35.0                         NaN               NaN   
5                NaN                         2.0               1.0   
6               16.0                         1.0               1.0   
7                NaN                         3.0               NaN   
8               36.0                         3.0               4.0   
9                NaN                         1.0               NaN   
10              12.0                         3.0               3.0   
11              52.0                         2.0               NaN   
12              12.0                         2.0               1.0   
13              28.0                         1.0               1.0   
14              28.0                         3.0               1.0   
15              48.0                         3.0               1.0   
16              36.0                         3.0               3.0   
17              24.0                         NaN               NaN   
18              21.0                         1.0               1.0   
19              35.0                         4.0               3.0   

    mucous membranes  capillary refill time  ...  packed cell volume  \
0                NaN                    2.0  ...                45.0   
1                4.0                    1.0  ...                50.0   
2                3.0                    1.0  ...                33.0   
3                6.0                    2.0  ...                48.0   
4                6.0                    2.0  ...                74.0   
5                3.0                    1.0  ...                 NaN   
6                1.0                    1.0  ...                37.0   
7                NaN                    1.0  ...                44.0   
8                3.0                    1.0  ...                38.0   
9                1.0                    1.0  ...                40.0   
10               5.0                    1.0  ...                44.0   
11               2.0                    1.0  ...                50.0   
12               1.0                    1.0  ...                 NaN   
13               2.0                    1.0  ...                37.0   
14               1.0                    1.0  ...                46.0   
15               4.0                

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   Age                          300 non-null    int64  
 2   Hospital Number              300 non-null    int64  
 3   rectal temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory rate             242 non-null    float64
 6   temperature of extremities   244 non-null    float64
 7   peripheral pulse             231 non-null    float64
 8   mucous membranes             253 non-null    float64
 9   capillary refill time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal distension         244 non-null    float64
 13  nasogastric tube    

В датафрейме 28 колонок и 300 рядов. Особи были обследованы очень подробно(28 колонок = признаков). 

В 1-й таблице бросаются в глаза неправильные данные в Возрасте - 9 вместо 1 или 2(ключ в описании). Это придется уточнять. Возможно, что просто 9 вместо 2, но нужно проверить.

Видим, что некоторые колонки содержат много пропусков. Возможно, данные просто не снимались, в зависимости от болезни лошади, или по другим причинам. Поэтому будем выбирать колонки, важные на наш взгляд. 

Также отсутствуют данные в некоторых колонках (? и 0 (в тип поражения2 и 3, согласно ключу)). Зависит от задачи, будем ли уточнять эти данные или исключим вместе с колонками. В развернутом файле, в конце в колонках тип поражения2 и 3 также нули. Учитывая, что нам нужно отобрать только 8 колонок, то эти с нулями точно исключим.

Категориальные данные вроде Номера больницы вряд ли нам будут полезны, тоже удаляем.

Оставим для дальнейшего анализа пока 12 колонок, самых важных на мой взгляд, чтобы ничего не упустить.

Отфильтруем колонки:

In [47]:
filtr_columns = df.filter(items = ['surgery', 'Age', 'rectal temperature', 'pulse', 'mucous membranes', 'pain', 'peristalsis', 'abdominal distension', 'abdomen', 'outcome', 'surgical lesion',  'type of lesion1'])
filtr_columns.head(30)

surgery  Age  rectal temperature  pulse  mucous membranes  pain  \
0       2.0    1                38.5   66.0               NaN   5.0   
1       1.0    1                39.2   88.0               4.0   3.0   
2       2.0    1                38.3   40.0               3.0   3.0   
3       1.0    9                39.1  164.0               6.0   2.0   
4       2.0    1                37.3  104.0               6.0   NaN   
5       2.0    1                 NaN    NaN               3.0   2.0   
6       1.0    1                37.9   48.0               1.0   3.0   
7       1.0    1                 NaN   60.0               NaN   NaN   
8       2.0    1                 NaN   80.0               3.0   4.0   
9       2.0    9                38.3   90.0               1.0   5.0   
10      1.0    1                38.1   66.0               5.0   3.0   
11      2.0    1                39.1   72.0               2.0   2.0   
12      1.0    1                37.2   42.0               1.0   3.0   
13      2.0    9                38.0   92.0               2.0   1.0   
14      1.0    1                38.2   76.0               1.0   3.0   
15      1.0    1                37.6   96.0               4.0   5.0   
16      1.0    9                 NaN  128.0               4.0   4.0   
17      2.0    1                37.5   48.0               NaN   NaN   
18      1.0    1                37.6   64.0               2.0   2.0   
19      2.0    1                39.4  110.0               6.0   NaN   
20      1.0    1                39.9   72.0               5.0   5.0   
21      2.0    1                38.4   48.0               1.0   1.0   
22      1.0    1                38.6   42.0               4.0   2.0   
23      1.0    9                38.3  130.0               NaN   2.0   
24      1.0    1                38.1   60.0               3.0   NaN   
25      2.0    1                37.8   60.0               NaN   NaN   
26      1.0    1                38.3   72.0               3.0   3.0   
27      1.0    1                37.8   48.0               1.0   NaN   
28      1.0    1                 NaN    NaN               NaN   NaN   
29      2.0    1                37.7   48.0               1.0   1.0   

    peristalsis  abdominal distension  abdomen  outcome  surgical lesion  \
0           4.0                   4.0      5.0      2.0                2   
1           4.0                   2.0      2.0      3.0                2   
2           3.0                   1.0      1.0      1.0                2   
3           4.0                   4.0      NaN      2.0                1   
4           NaN                   NaN      NaN      2.0                2   
5           3.0                   2.0      3.0      1.0                2   
6           3.0                   3.0      5.0      1.0                1   
7           4.0                   2.0      4.0      2.0                1   
8           4.0                   4.0      5.0      3.0                1   
9           3.0                   1.0      NaN      1.0                2   
10          3.0                   1.0      5.0      1.0                1   
11          1.0                   2.0      4.0      1.0                1   
12          3.0                   3.0      5.0      1.0                2   
13          3.0                   2.0      1.0      2.0                2   
14          4.0                   1.0      4.0      1.0                1   
15          3.0                   3.0      NaN      2.0                1   
16          4.0                   3.0      5.0      2.0                2   
17          NaN                   NaN      NaN      1.0                2   
18          3.0                   1.0      5.0      1.0                1   
19          3.0                   3.0      NaN      1.0                2   
20          4.0                   4.0      4.0      1.0                1   
21          3.0                   1.0      3.0      1.0                2   
22          3.0                   1.0      NaN      1.

In [48]:
filtr_columns.describe()

surgery         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       mucous membranes       pain  peristalsis  abdominal distension  \
count        253.000000  245.00000   256.000000            244.000000   
mean           2.853755    2.95102     2.917969              2.266393   
std            1.620294    1.30794     0.976744              1.065131   
min            1.000000    1.00000     1.000000              1.000000   
25%            1.000000    2.00000     3.000000              1.000000   
50%            3.000000    3.00000     3.000000              2.000000   
75%            4.000000    4.00000     4.000000              3.000000   
max            6.000000    5.00000     4.000000              4.000000   

          abdomen     outcome  surgical lesion  type of lesion1  
count  182.000000  299.000000       300.000000       300.000000  
mean     3.692308    1.551839         1.363333      3657.880000  
std      1.491725    0.737187         0.481763      5399.513513  
min      1.000000    1.000000         1.000000         0.000000  
25%      2.000000    1.000000         1.000000      2111.750000  
50%      4.000000    1.000000         1.000000      2673.500000  
75%      5.000000    2.000000         2.000000      3209.000000  
max      5.000000    3.000000         2.000000     41110.000000

Здесь большинство значений - категории, выраженные в виде чисел. Поэтому подсчеты среднего и т.п. к ним не применимы.

Видно, что максимальный возраст - 9. Нужно проверить, чтобы не было других чисел(напр, 2), кроме 1 и 9 и уточнить, действительно ли 9 - это молодая лошадь до полугода.

Колонки с категориальными данными в этой отфильтрованной выборке:
Все, кроме rectal temperature и pulse

Поэтому далее будем считать числовые.

Найдем пограничные значения.

In [49]:
print(max(filtr_columns['rectal temperature']))
print(min(filtr_columns['rectal temperature']))
temp_range = filtr_columns['rectal temperature'].max() - filtr_columns['rectal temperature'].min()
print(temp_range)               #Размах температуры вроде реальный, у больных темп может быть 35.4 и тем более 40.8

40.8
35.4
5.399999999999999


In [50]:
print(max(filtr_columns['pulse']))
print(min(filtr_columns['pulse']))
pulse_range = filtr_columns['pulse'].max() - filtr_columns['pulse'].min()
print(pulse_range)                

184.0
30.0
154.0


In [51]:
print(filtr_columns['rectal temperature'].std())
print(filtr_columns['pulse'].std())

0.7322886641121578
28.630556660735003


СКО температуры очень маленькое и можно заменить средним. А с пульсом так нельзя, слишком большое СКО, будем заменять медианой по категориям молодых и старых лошадей.

In [52]:
print(filtr_columns['rectal temperature'].mean())

38.16791666666669


In [54]:
print(filtr_columns['pulse'].median())
print(filtr_columns['pulse'].mode()[0])

64.0
48.0


In [55]:
fill_rectemp = filtr_columns
fill_rectemp['rectal temperature'].fillna(filtr_columns['rectal temperature'].mean(), inplace=True)
fill_rectemp['rectal temperature'].fillna(filtr_columns['rectal temperature'].mean()).isna().sum()
print(fill_rectemp['rectal temperature'].describe())

count    300.000000
mean      38.167917
std        0.654705
min       35.400000
25%       37.900000
50%       38.167917
75%       38.500000
max       40.800000
Name: rectal temperature, dtype: float64


CKO незначительно уменьшилось - на 0,08. Подходит.

In [56]:
filtr_columns['pulse'].fillna(filtr_columns['pulse'].median(), inplace=True)
filtr_columns['pulse'].fillna(filtr_columns['pulse'].median()).isna().sum()
print(filtr_columns['pulse'].describe())

count    300.000000
mean      71.280000
std       27.541545
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64


In [57]:
print(filtr_columns.groupby('Age')['pulse'].median())
fill_median_by_age = filtr_columns
fill_median_by_age['pulse'].fillna(filtr_columns.groupby('Age')['pulse'].transform('median'), inplace=True)

Age
1     64.0
9    122.0
Name: pulse, dtype: float64


Группировка по возрасту оправдала наше предположение о разности пульсов в зависимости от возраста. А также то, что есть ошибка в описании и "2" в описании соответствует "9" в датафрейме. Но пульс высоковат, что дополнительно указывает на то, что выборка содержит данные по проблемным лошадям.
Поэтому посмотрим зависимость от болезни.

In [58]:
print(filtr_columns.groupby(['Age', 'surgical lesion'])['pulse'].median())
fill_median_by_age = filtr_columns
fill_median_by_age['pulse'].fillna(filtr_columns.groupby(['Age', 'surgical lesion'])['pulse'].transform('median'), inplace=True)

Age  surgical lesion
1    1                   66.0
     2                   52.0
9    1                  130.0
     2                   92.0
Name: pulse, dtype: float64


Видим зависимость пульса от хирургического вмешательства (1 - Да). Если добавить исход (выжила ли лошадь), то получим:

In [59]:
print(filtr_columns.groupby(['Age', 'outcome', 'surgical lesion'])['pulse'].median())
fill_median_by_age = filtr_columns
fill_median_by_age['pulse'].fillna(filtr_columns.groupby(['Age', 'outcome', 'surgical lesion'])['pulse'].transform('median'), inplace=True)

Age  outcome  surgical lesion
1    1.0      1                   64.0
              2                   48.0
     2.0      1                   72.0
              2                   75.0
     3.0      1                   72.0
              2                   88.0
9    1.0      1                  124.0
              2                   90.0
     2.0      1                  146.0
              2                  110.0
     3.0      1                  100.0
Name: pulse, dtype: float64


outcome: 1 - выжила, 2 - умерла, 3 - усыплена.
Видим, что хирург.вмешательство особенно повлияло на пульс у выживших взрослых и молодых особей, а у молодняка еще у умерших (у него норма 100-120), но поскольку они всё равно умерли, то здесь нужно искать другие причинно-следственные связи.
Оставим последнюю схему замены НаНов в колонке Пульс - она показывает больше корреляций.

In [60]:
filtr_columns.info()    #Проверка подмен в колонках Пульс и рект.температура. Заменено

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      253 non-null    float64
 5   pain                  245 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               182 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


Очень много нанов в abdomen. Просматривая данные в датафрейме, видим некоторую зависимость между abdomen и surgical lesion.

In [61]:
print(filtr_columns.groupby('surgical lesion')['abdomen'].median())
fill_median_by_surgical = filtr_columns
fill_median_by_surgical['abdomen'].fillna(filtr_columns.groupby('surgical lesion')['abdomen'].transform('median'), inplace=True)

surgical lesion
1    5.0
2    2.0
Name: abdomen, dtype: float64


In [62]:
filtr_columns.info()    #Заменено + Проверка = Подходит (среднее изменилось на 0,04 и СКО на 0,01):

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      253 non-null    float64
 5   pain                  245 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


In [63]:
print(df['abdomen'].describe())

count    182.000000
mean       3.692308
std        1.491725
min        1.000000
25%        2.000000
50%        4.000000
75%        5.000000
max        5.000000
Name: abdomen, dtype: float64


In [64]:
print(fill_median_by_surgical['abdomen'].describe())

count    300.000000
mean       3.736667
std        1.483687
min        1.000000
25%        2.000000
50%        4.500000
75%        5.000000
max        5.000000
Name: abdomen, dtype: float64


# Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [20]:
q1 = filtr_columns['pulse'].quantile(0.25)
q3 = filtr_columns['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = filtr_columns[filtr_columns['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('Age')
remove_outliers

surgery  Age  rectal temperature  pulse  mucous membranes  pain  \
0        2.0    1           38.500000   66.0               NaN   5.0   
198      2.0    1           38.300000   60.0               1.0   1.0   
197      2.0    1           38.500000   96.0               NaN   2.0   
196      2.0    1           38.500000   60.0               5.0   2.0   
195      2.0    1           38.000000   44.0               1.0   NaN   
..       ...  ...                 ...    ...               ...   ...   
252      1.0    9           38.800000  124.0               2.0   2.0   
191      2.0    9           38.167917  100.0               1.0   4.0   
23       1.0    9           38.300000  130.0               NaN   2.0   
287      2.0    9           39.500000   84.0               NaN   NaN   
39       1.0    9           39.200000  146.0               NaN   NaN   

     peristalsis  abdominal distension  abdomen  outcome  surgical lesion  \
0            4.0                   4.0      5.0      2.0                2   
198          3.0                   1.0      2.0      1.0                2   
197          4.0                   2.0      5.0      2.0                1   
196          2.0                   1.0      3.0      3.0                2   
195          1.0                   2.0      2.0      1.0                2   
..           ...                   ...      ...      ...              ...   
252          3.0                   4.0      4.0      2.0                1   
191          1.0                   1.0      2.0      1.0                2   
23           4.0                   NaN      5.0      1.0                1   
287          NaN                   NaN      2.0      1.0                2   
39           NaN                   NaN      5.0      2.0                1   

     type of lesion1  
0              11300  
198                0  
197             1400  
196              300  
195                0  
..               ...  
252             2208  
191             7111  
23              3111  
287                0  
39              2113  

[295 rows x 12 columns]

In [21]:
pd.concat([filtr_columns, remove_outliers]).drop_duplicates(keep=False)

surgery  Age  rectal temperature  pulse  mucous membranes  pain  \
3        1.0    9                39.1  164.0               6.0   2.0   
41       2.0    9                39.0  150.0               NaN   NaN   
55       1.0    9                38.6  160.0               5.0   3.0   
255      1.0    9                38.8  184.0               1.0   4.0   
275      1.0    9                38.8  150.0               6.0   5.0   

     peristalsis  abdominal distension  abdomen  outcome  surgical lesion  \
3            4.0                   4.0      5.0      2.0                1   
41           NaN                   NaN      5.0      1.0                1   
55           3.0                   4.0      5.0      2.0                1   
255          1.0                   3.0      5.0      2.0                1   
275          3.0                   2.0      5.0      2.0                1   

     type of lesion1  
3               2208  
41              9400  
55              7111  
255             7111  
275             4207

Найдено 5 выбросов. 
184 - с большей вероятностью, неправильный ввод данных - уточнить у замерщика. Такое значение едва ли реально, в лит-ре встречается максимальный пульс 175, та и то во время бега. 
Остальные значения могут быть причиной:
1) неправильного измерения (во время бега) - взять методику измерений, расспросить замерщиков, узнать, тренируют ли лошадей до полугода
2) ошибкой ввода данных
3) маячком большой патологии. Мы видим, что все они подвергались хирургическому вмешательству и умерли все из них, кроме одной, которая не подвергалась операции (тип вмешательства - мочевой пузырь). Учитывая, что все особи относятся к молодняку, у которого средний пульс 90-120, то пульс 150 для патологий вполне допустим.

Таким образом, сначала соберем сведения и узнаем причины выбросов, а если такой возможности нет, то стоит заменить выбросы на 150 как максимальный для жеребцов. Потому что для патологий пульс 150 более чем приемлем в состоянии покоя.

In [22]:
print(remove_outliers['pulse'].mean())
print(remove_outliers['pulse'].median())

69.74915254237288
64.0


In [23]:
print(filtr_columns['pulse'].mean() - remove_outliers['pulse'].mean())

1.530847457627118


In [24]:
print(filtr_columns['pulse'].median() - remove_outliers['pulse'].median())

0.0


# Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

Продолжим работать с пропусками. У нас осталось 4 столбца с нанами: mucous membranes, pain,	peristalsis, abdominal distension

In [25]:
filtr_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      253 non-null    float64
 5   pain                  245 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


С этими категориями всё сложно, потому что трудно найти зависимости. Просмотрим данные.

In [26]:
filtr_columns['mucous membranes'].value_counts()

1.0    79
3.0    58
4.0    41
2.0    30
5.0    25
6.0    20
Name: mucous membranes, dtype: int64

In [27]:
print(filtr_columns['mucous membranes'].median())
print(filtr_columns['mucous membranes'].mode()[0])

3.0
1.0


- 1 and 2 probably indicate a normal or slightly increased circulation
- 3 may occur in early shock
Таким образом, ни 1, ни 3 не является критическим значением, тем более, что значение 3 является вероятностным.
Поэтому заменим наны тройкой - медианой.
Такой разброс значений (выше) указывает, что критических значений немного, а значит, не имеет прямой связи с болезнями. И это доказывает:

In [28]:
print(filtr_columns.groupby('surgical lesion')['mucous membranes'].median())
fill_median_by_membra = filtr_columns
fill_median_by_membra['mucous membranes'].fillna(filtr_columns.groupby('surgical lesion')['mucous membranes'].transform('median'), inplace=True)

surgical lesion
1    3.0
2    2.0
Name: mucous membranes, dtype: float64


In [30]:
filtr_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      300 non-null    float64
 5   pain                  245 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


Для боли попробуем найти зависимость с проблемным животом:

In [31]:
print(filtr_columns.groupby('abdomen')['pain'].median())
fill_median_by_pain = filtr_columns
fill_median_by_pain['pain'].fillna(filtr_columns.groupby('abdomen')['pain'].transform('median'), inplace=True)

abdomen
1.0    2.0
2.0    2.0
3.0    2.5
4.0    3.0
5.0    3.0
Name: pain, dtype: float64


pain - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
abdomen
          - possible values
               1 = normal
               2 = other

pain:     3 = intermittent mild pain
              
abdomen   - 3 is probably an obstruction caused by a mechanical impaction
            and is normally treated medically
          - 4 and 5 indicate a surgical lesion 


In [32]:
filtr_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      300 non-null    float64
 5   pain                  300 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


Аналогично для перистальтики кишечника и вздутие живота. Логично предположить зависимость от живота, но не подтвердилось. Учитвая, что это не особо важный показатель, заменим модой или медианой, тем более что они совпадают.

In [36]:
print(filtr_columns.groupby('abdomen')['peristalsis'].median())
fill_median_by_peristalsis = filtr_columns
fill_median_by_peristalsis['peristalsis'].fillna(filtr_columns.groupby('abdomen')['peristalsis'].transform('median'), inplace=True)

abdomen
1.0    2.0
2.0    3.0
3.0    3.0
4.0    3.0
5.0    3.0
Name: peristalsis, dtype: float64


In [38]:
filtr_columns['peristalsis'].value_counts()

3.0    170
4.0     73
1.0     39
2.0     18
Name: peristalsis, dtype: int64

In [39]:
print(filtr_columns['peristalsis'].median())
print(filtr_columns['peristalsis'].mode()[0])

3.0
3.0


In [40]:
fill_mode_peristalsis = filtr_columns
fill_mode_peristalsis['peristalsis'].fillna(filtr_columns['peristalsis'].mode()[0], inplace=True)

fill_mode_peristalsis['peristalsis'].value_counts()

3.0    170
4.0     73
1.0     39
2.0     18
Name: peristalsis, dtype: int64

In [41]:
filtr_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      300 non-null    float64
 5   pain                  300 non-null    float64
 6   peristalsis           300 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB


Аналогично для abdominal distension. Здесь больше подойдет медиана:

In [42]:
filtr_columns['abdominal distension'].value_counts()

1.0    76
3.0    65
2.0    65
4.0    38
Name: abdominal distension, dtype: int64

In [43]:
print(filtr_columns['abdominal distension'].median())
print(filtr_columns['abdominal distension'].mode()[0])

2.0
1.0


In [44]:
fill_median_distension = filtr_columns
fill_median_distension['abdominal distension'].fillna(filtr_columns['abdominal distension'].median(), inplace=True)
fill_median_distension['abdominal distension'].fillna(filtr_columns['abdominal distension'].median()).isna().sum()

0

In [45]:
print(fill_median_distension['abdominal distension'].describe())

count    300.000000
mean       2.216667
std        0.965832
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        4.000000
Name: abdominal distension, dtype: float64


In [65]:
filtr_columns.info()            #Все наны заменены.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               299 non-null    float64
 1   Age                   300 non-null    int64  
 2   rectal temperature    300 non-null    float64
 3   pulse                 300 non-null    float64
 4   mucous membranes      253 non-null    float64
 5   pain                  245 non-null    float64
 6   peristalsis           256 non-null    float64
 7   abdominal distension  244 non-null    float64
 8   abdomen               300 non-null    float64
 9   outcome               299 non-null    float64
 10  surgical lesion       300 non-null    int64  
 11  type of lesion1       300 non-null    int64  
dtypes: float64(9), int64(3)
memory usage: 28.2 KB
